redhat-appstudio/service-provider-integration-operator
# feat: enable oauth flow for remotesecrets

### What does this PR do?
- adds a new controller that watches RemoteSecrets that have label `appstudio.redhat.com/sp.host`, and if the service provider with such host supports OAuth, it adds annotation `appstudio.redhat.com/sp.oauthurl` with OAuth URL.
- refactor OAuth service to support OAuth from both SPIAccessToken and RemoteSecret at the same time. This involves changing the format of OAuth URL state parameter.

### Potential incompatibilities
- new controller watching RemoteSecret - it only patches annotations and conditions. There should not be any major conflict with the RemoteSecret operator
- the format of the `state` query parameter of the OAuth URL is changed - the only way this change may produce issues is when the user is going through OAuth flow on SPIAccessToken, is between /authenticate and /callback request, and the SPI operator is updated to version containing this change (which is **very** unlikely). However, if that happens, the user just needs to restart the OAuth flow from the new OAuth URL.


### Security
To go through the OAuth flow, the user must have  `get` and `update` rbac permissions for remotesecrets.

I tested the PR on Openshift to ensure the permissions are checked.

What the user sees in browser if he does not have permissions:
![image](https://github.com/redhat-appstudio/service-provider-integration-operator/assets/73115616/20e1cde8-94ab-488e-9242-60d99be65049)
OAuth service logs:
![image](https://github.com/redhat-appstudio/service-provider-integration-operator/assets/73115616/d3f7d545-a444-4b26-a83c-4876f1b68561)


### Screenshot/screencast of this PR



### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-566

### How to test this PR?

#### Complete GitHub OAuth flow 
1. Deploy SPI (operator: `quay.io/pbaran/spio:[SVPI-566](https://issues.redhat.com//browse/SVPI-566)-oauth-simple` service: `quay.io/pbaran/spis:[SVPI-566](https://issues.redhat.com//browse/SVPI-566)-oauth-simple`) and RemoteSecret operator (from main) - may need to update the RBAC rules for SPI operator so that it can access RemoteSecrets.
2. Setup OAuth on GitHub with SPI, update SPI provider configuration.
3. Create RemoteSecret with github host label:
```shell
cat <<EOF | kubectl apply  -f -
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  namespace: default
  name: test-rs
  labels:
    appstudio.redhat.com/sp.host: github.com
spec:
  secret:
    generateName: some-secret-
  targets:
    - namespace: default
EOF
```
4. check and copy the OAuth URL from RS annotations:
![image](https://github.com/redhat-appstudio/service-provider-integration-operator/assets/73115616/56a5d319-0d6e-4f6d-858d-1d3e424d8e9a)
5. Get a service account token that can `get` and `update` RemoteSecrets - on minikube it should not be necessary (hack/get-default-sa-token.sh).  
6. Use /login endpoint of the OAuth Service with the token to setup a cookie session. (either with `hack/oauth-ui.html` or other means)
7. Open the OAuth URL in browser, go through the OAuth flow.
8. Verify that RemoteSecret has data and target Secret was created and has expected data:
![image](https://github.com/redhat-appstudio/service-provider-integration-operator/assets/73115616/b800c5b6-e8a5-4947-94c6-8596de6464a0)

#### Security and other test cases
TBD
```
quay.io/redhat-appstudio/pull-request-builds:spi-controller-#
quay.io/redhat-appstudio/pull-request-builds:spi-oauth-#
```
